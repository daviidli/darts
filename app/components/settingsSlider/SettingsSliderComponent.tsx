import React, { useState, useEffect } from 'react';
import { Typography, Slider } from '@material-ui/core';

export type Props = {
	clearDartsWaitTime: number;
	setWaitTime: (wait: number) => {};
	className?: string;
};

const SettingsSliderComponent = (props: Props) => {
	const { clearDartsWaitTime, setWaitTime, className } = props;
	const [sliderValue, setSliderValue] = useState<number>(clearDartsWaitTime);

	useEffect(() => {
		setSliderValue(clearDartsWaitTime);
	}, [clearDartsWaitTime, setSliderValue]);

	const marks = [
		{
			value: 0,
			label: 'âˆž'
		},
		{
			value: 30,
			label: '30s'
		},
		{
			value: 60,
			label: '60s'
		},
		{
			value: 90,
			label: '90s'
		},
		{
			value: 120,
			label: '120s'
		}
	];

	return (
		<div className={className}>
			<Typography gutterBottom>Clear Darts Timeout</Typography>
			<Slider
				value={sliderValue}
				min={0}
				max={120}
				step={5}
				valueLabelDisplay="auto"
				marks={marks}
				onChange={(_e, val) => setSliderValue(val as number)}
				onChangeCommitted={(_e, val) => {
					setWaitTime(val as number);
				}}
			/>
		</div>
	);
};

export default SettingsSliderComponent;
