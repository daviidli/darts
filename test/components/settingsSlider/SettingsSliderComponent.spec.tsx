import { spy } from 'sinon';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import SettingsSliderComponent, {
	Props
} from '../../../app/components/settingsSlider/SettingsSliderComponent';

Enzyme.configure({ adapter: new Adapter() });

// Mock material-ui's slider component
// The functionality of the slider doesn't need to be tested, but the onChange
// and onChangeCommitted functions should be.
// These functions are wrapped in the mocked component because enzyme doesn't send
// multiple args to mounted components.
jest.mock('@material-ui/core/Slider', () => props => {
	const { value, min, max, step, marks, onChange, onChangeCommitted } = props;
	const handleChange = e => onChange(e, e.mockValue);
	const handleChangeCommitted = e => onChangeCommitted(e, e.mockValue);
	return (
		<>
			<div className="mockSlider" {...{ value, min, max, step, marks }}>
				{value}
			</div>
			<div className="mockSliderOnChange" onChange={handleChange} />
			<div
				className="mockSliderOnChangeCommitted"
				onChange={handleChangeCommitted}
			/>
		</>
	);
});

const initialProps: Props = {
	clearDartsWaitTime: 0,
	setWaitTime: () => ({})
};

const setup = (props: Props = initialProps) => {
	const wrapper = mount(<SettingsSliderComponent {...props} />);
	return {
		wrapper,
		slider: wrapper.find('.mockSlider'),
		sliderOnChange: wrapper.find('.mockSliderOnChange'),
		sliderOnChangeCommitted: wrapper.find('.mockSliderOnChangeCommitted')
	};
};

describe('Settings Slider component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should have correct initial value', () => {
		const props = { ...initialProps };
		props.clearDartsWaitTime = 30;
		const { slider } = setup(props);
		expect(slider.text()).toBe('30');
	});

	it('should update slider value on props update', () => {
		const { wrapper, slider } = setup();
		expect(slider.text()).toBe('0');
		wrapper.setProps({ clearDartsWaitTime: 5 });
		return new Promise(resolve => {
			setImmediate(() => {
				expect(slider.text()).toMatch('5');
				resolve();
			});
		});
	});

	it('should set slider value on change event', () => {
		const setWaitTimeMock = spy();
		const props = { ...initialProps };
		props.setWaitTime = setWaitTimeMock;
		const { slider, sliderOnChange } = setup(props);
		expect(slider.text()).toMatch('0');
		sliderOnChange.simulate('change', { mockValue: 5 });
		expect(slider.text()).toMatch('5');
		expect(setWaitTimeMock.notCalled).toBeTruthy();
	});

	it('should set wait time on change committed event', () => {
		const setWaitTimeMock = spy();
		const props = { ...initialProps };
		props.setWaitTime = setWaitTimeMock;
		const { sliderOnChangeCommitted } = setup(props);
		sliderOnChangeCommitted.simulate('change', { mockValue: 30 });
		expect(setWaitTimeMock.calledOnce).toBeTruthy();
		expect(setWaitTimeMock.calledWith(30)).toBeTruthy();
	});
});
