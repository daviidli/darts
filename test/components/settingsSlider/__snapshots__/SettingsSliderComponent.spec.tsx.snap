// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Slider component should match snapshot 1`] = `
<SettingsSliderComponent
  clearDartsWaitTime={0}
  setWaitTime={[Function]}
>
  <div>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
    >
      <ForwardRef(Typography)
        classes={
          Object {
            "alignCenter": "MuiTypography-alignCenter",
            "alignJustify": "MuiTypography-alignJustify",
            "alignLeft": "MuiTypography-alignLeft",
            "alignRight": "MuiTypography-alignRight",
            "body1": "MuiTypography-body1",
            "body2": "MuiTypography-body2",
            "button": "MuiTypography-button",
            "caption": "MuiTypography-caption",
            "colorError": "MuiTypography-colorError",
            "colorInherit": "MuiTypography-colorInherit",
            "colorPrimary": "MuiTypography-colorPrimary",
            "colorSecondary": "MuiTypography-colorSecondary",
            "colorTextPrimary": "MuiTypography-colorTextPrimary",
            "colorTextSecondary": "MuiTypography-colorTextSecondary",
            "displayBlock": "MuiTypography-displayBlock",
            "displayInline": "MuiTypography-displayInline",
            "gutterBottom": "MuiTypography-gutterBottom",
            "h1": "MuiTypography-h1",
            "h2": "MuiTypography-h2",
            "h3": "MuiTypography-h3",
            "h4": "MuiTypography-h4",
            "h5": "MuiTypography-h5",
            "h6": "MuiTypography-h6",
            "noWrap": "MuiTypography-noWrap",
            "overline": "MuiTypography-overline",
            "paragraph": "MuiTypography-paragraph",
            "root": "MuiTypography-root",
            "srOnly": "MuiTypography-srOnly",
            "subtitle1": "MuiTypography-subtitle1",
            "subtitle2": "MuiTypography-subtitle2",
          }
        }
        gutterBottom={true}
      >
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          Clear Darts Timeout
        </p>
      </ForwardRef(Typography)>
    </WithStyles(ForwardRef(Typography))>
    <Component
      marks={
        Array [
          Object {
            "label": "∞",
            "value": 0,
          },
          Object {
            "label": "30s",
            "value": 30,
          },
          Object {
            "label": "60s",
            "value": 60,
          },
          Object {
            "label": "90s",
            "value": 90,
          },
          Object {
            "label": "120s",
            "value": 120,
          },
        ]
      }
      max={120}
      min={0}
      onChange={[Function]}
      onChangeCommitted={[Function]}
      step={5}
      value={0}
      valueLabelDisplay="auto"
    >
      <div
        className="mockSlider"
        marks={
          Array [
            Object {
              "label": "∞",
              "value": 0,
            },
            Object {
              "label": "30s",
              "value": 30,
            },
            Object {
              "label": "60s",
              "value": 60,
            },
            Object {
              "label": "90s",
              "value": 90,
            },
            Object {
              "label": "120s",
              "value": 120,
            },
          ]
        }
        max={120}
        min={0}
        step={5}
        value={0}
      >
        0
      </div>
      <div
        className="mockSliderOnChange"
        onChange={[Function]}
      />
      <div
        className="mockSliderOnChangeCommitted"
        onChange={[Function]}
      />
    </Component>
  </div>
</SettingsSliderComponent>
`;
